sky/backends/backend_utils.py

```diff
@@ -960,6 +959,7 @@ def write_cluster_config(
 
                 # GCP only:
                 'gcp_project_id': gcp_project_id,
+                'specific_reservations': to_provision.specific_reservations,
```

Comment:
Question: Should we check if reserved resources are available before passing this argument to the ray yaml file?

In other words, will GCP create normal on-demand instance if this `specific_reservations` is specified, and all the reserved resources are occupied?

If all the reserved instances are occupied, and the optimizer calculated the cost by the on-demand instances, we should provision an on-demand instance here, instead of using the specific reservations? Otherwise, we will never be able to use normal on-demand instances?

Reply:
I'm also concerned about potential race conditions if we have to manually add or remove parameters. Consider a scenario where a user initiates the same task twice, each time with different parameters. There's a risk that the parameter for the second task might not be removed. This could occur if the first task reserves the resource shortly after the second task checks the reservation status. As a result, the second task could fail during provisioning because it's operating under the incorrect assumption that the resource is available.

Reply:
Oops, it seems I replied to the wrong place. copying it here (https://github.com/skypilot-org/skypilot/pull/2352#discussion_r1286033091)

> I also fear that it would fail if the reservation had, lets say, 1 available resource but was requiring 2.
I will properly test this today and try to find a solution.

As we mentioned in the [doc](https://docs.google.com/document/u/1/d/1STKlU14RgLxvyMPzpaMeVcm6Dswz8ShJAZtGM2uLcsM/edit?ouid=100549593362412474252&usp=docs_home&ths=true), maybe we can do the following `gcp-ray.YAML`, i.e., we only request for the number of reserved instances that is available. 

Note: the `num_reserved_nodes` may need to be make the same as the value used in optimizer, so we make sure the cluster actually launched has the same cost as estimated by the optimizer.

```yaml

# gcp-ray.yaml

ray_head_default:
      …
      {%- if num_reserved_nodes > 1 %}
      reservationAffinity:
        consumeReservationType: SPECIFIC_RESERVATION
        key: compute.googleapis.com/reservation-name
        values: {{ specific_reservations }}
      {%- endif %}


{%- if num_reserved_nodes > 1 %}
ray_worker_reserved:
    min_workers: {{ num_reserved_nodes - 1 }}
    max_workers: {{ num_reserved_nodes - 1 }}
    …
    reservationAffinity:
        consumeReservationType: SPECIFIC_RESERVATION
        key: compute.googleapis.com/reservation-name
        values: {{ specific_reservations }}
 {%- endif %}

{%- if num_nodes - num_reserved_nodes > 1 %}
ray_worker_default:
    min_workers: {{ num_nodes - num_reserved_nodes - 1 }}
    max_workers: {{ num_nods - num_reserved_nodes - 1 }}
    …
{%- endif %}
```

> I'm also concerned about potential race conditions if we have to manually add or remove parameters. Consider a scenario where a user initiates the same task twice, each time with different parameters. There's a risk that the parameter for the second task might not be removed. This could occur if the first task reserves the resource shortly after the second task checks the reservation status. As a result, the second task could fail during provisioning because it's operating under the incorrect assumption that the resource is available.

Yes, that is a limitation, but I guess it should be fine as a first step to fail the provisioning if the race condition happens, as our higher-level failover will then go to the next zone/region to find the available resources for the task. It might be sub-optimal, but when the `--retry-until-up` is provided and no available resources is available in other locations, the failover will go to the failed zone again and try to provision the cluster with the correct number of reserved instances. In the future, we can probably change the logic for the `task._blocked_resources` and optimizer a bit, to have the failover retry the same zone a few times before failover to the next, if reserved instances are failed to be launched.